README.md
# 004

Saccadic landing positions on isolated photographs of daily-life graspable objects as a function of handle side.
In revision for JoV.

## analysis

The goal is to have analysis scripts that work for all 3 experiments. I started to make scripts for 004C. Now, I'm adapting them such that they work for 004B as well. Todo: adapt for 004A (with fixation checks etc.)


### parse.py()

- From EyeLink events to dm containing raw fixation coordinates and time stamps.
- We use saccade events instead of fixation events. This is because it often occurred that a fixation started but not ended (before the stop-recording item), which is why we loose a lot of trials.
- If sacc_ey (y coordinate LP) is too close to the center (to the fix dot), the fixation is skipped (not used for the analysis).

### bbox.py()

- Determines the width and height of a bbox around the non-white pixels in the PNG file.
- We need these new width and height to normalize LPs (see [normOnWidth.py] relative to the actual stimulus (rather than the w and h of the PNG file)

### getDm.py()

Converts raw coordinates to meaningful coordinates:

- [centralOrigin.py] Normalize such that origin = (0,0)
- [rotate.py] Rotate as if object was presented at -90 in UVF
- [flip.py] Normalize on orientation, as if handle was always to the right
- [normOnWidth.py] Normalize on object width
	
### centralOrigin.py

- From OpenSesame coordinates (origin at top left) to normalized coordinates (origin at center)
- This is necessary to be able to rotate the landing positions

### rotate.py

- Rotates LPs as if object was always presented with a rotation angle of -90 degrees, that is, in the UPV with the CoG aligned on the vertical meridian
- For the formula, see: [http://en.wikipedia.org/wiki/Rotation_(mathematics)](http://en.wikipedia.org/wiki/Rotation_(mathematics))

### flip.py

- Flips landing positions if flip condition was "left".
- We can do this, because the LPs are already normalized on rotation (see [rotate.py])
- TODO: check whether this is ok

### normOnWidth.py

- Normalizes objects on object width, such that LPs become a percentile of the overall width.
- TODO: check with CoG



analysis for JoV v1/changelog.txt
CHANGELOG Françoise:

1. 
Reparsed raw data to determine not only the first saccade, but all saccades in the stimulus-response interval.
- For the first, on-object saccade I used the EyeLink criteria + an extra criterion: sacc > 2.6 degrees.
- For all subsequent within-object refixations I used the EyeLink criteria. 
- I excluded off-object fixations offline (0.25% for the first landing position, 0.83% for the second, 1.78 for the 
	third, and 2.78 for the fourth. TODO for Exp2)
The first three landing positions are quite easy to analyse. The fourth are more difficult, because
there are only 70 fourth saccades. For the second experiment (without fix checks) no more than 3 saccades
occurred.

The variables 'ToHandle' and 'ToContrast' are created during parsing. 

2. 
The CoG of all stimuli is calculated as follows:
- First, the original bitmap is filtered with a Sobel filter, which detects edges. We did this because
	edges are salient. TODO: does this mean color/saliency has no effect anymore?
- Next, we caclulated the CoG of those filtered stimuli by TODO.
If there are questions about this procedure: I admit that the choice is quite arbitrary, but the data from
Experiment 2 suggest that it worked pretty well. However, to systematically investigate which kind of CoG
calculation works best, an experiment with non-semantic stimuli could be run. Different CoG could be compared
to see which one explains most of the variance. 

TODO: make slide on calculation of CoG
TODO: show Figures presentation
TODO: show Figure containing all xCoG's per object

3. New variables are created, which contain the landing position corrected for CoG. More specifically, this
means that I subtracted the deviation between the bitmap's center and the CoG from the original landing position.
This is done for both horizontal and vertical coordinates, although we're most interested in horizontal landing
positions.

TODO: The results can be seen in the trial-by-trial plots.


4.
I normalised all original landing positions such that -.5 indicates the extreme left of the object and .5 indicates
the extreme right (irrespective of the exact width). Same for the vertical axes, but this is less useful
(since we're mainly interested in horizontal landing position) and more tricky (because of a huge 
undershoot -> TODO: show plots).

For the corrected EM data, 0 means that the eyes landed on the CoG. -.5 and and .5 do not mean the left and right side
of the object anymore.

TODO: show plots with dist hists.

5. Affordance effect:
RT can be determined in several ways:
- interval between stimulus onset and response
- interval between fix check succesfull and response
- interval between on-object landing and response
I chose the latter.
Exp 1: p = .10
Exp 2: n.s. (p = .48) TODO: Look closer into this

Filter criteria:
	- saccade too fast: < 80 ms TODO: ref? TODO: try without?
	- fix check 1 too long: > 700 ms
	- fix check 2 too long: > 450 ms TODO: where did I get those values from? Distributions?
	- non-handled objects (ruler, etc.)
	- error trials
	- RTs deviating more than 2.5 SD's from pp's condition mean are discarded

6. 
Effect of object orientation on landing position:
Filter criteria:
	- saccade too fast: < 80 ms TODO: ref? TODO: try without?
	- fix check 1 too long: > 700 ms
	- fix check 2 too long: > 450 ms TODO: where did I get those values from? Distributions?
	- only stimuli without contrast manipulation
	- only on-object landings (see above)
	- non-handled objects (ruler, etc.)
	- landing positions deviating more than 2.5 SD's from pp's condition mean are discarded

TODO: heavy side

6A:
Initial landing position as a function of object orientation:
	Exp1 original landing positions: away from handle (which is on average the heavy side
	Exp1 corrected landing positions: towards handle
	
	See landing position plots -> 
	- we shouldn't interpret this as bias towards the handle, because... TODO
	- tendency to move eyes as straight as possible
	- or: CoG is not calculated correctly and results in an over correction compared to 
		what is actually going on in the visual system.
	
	Exp2: no difference between both orientations. TODO: both don't differ from zero, i.e. the
	CoG and the center of the display (and NOT the center of the bitmap).
	
6B:
Refixations:
	No difference between experiments anymore. All refixations are directed towards the non-handled
	side.
	Interpretations:
		- Most informative (all objects are handled: to unique information (necessary for categorisation)
			is therefore on the other side).
		- The part performing the action (e.g. 'hammering', 'washing', etc., cf. Humphreys).
		- We cannot disociate between these and possible other explanations. The main point is that it seems
			that more high-level processing is guiding the eyes.

7:
What is the influence of initial landing position on object processing, as measured by refixation probability,
fixation duration, and RT? (Cf. OVP studies in words, and Henderson (1993), and Pajak & Nuthmann (2013) and 
Foulsham & Kingstone (2013), although the latter are studies on natural-scene viewing.

Filter criteria:
	- saccade too fast: < 80 ms TODO: ref? TODO: try without?
	- fix check 1 too long: > 700 ms
	- fix check 2 too long: > 450 ms TODO: where did I get those values from? Distributions?
	- only stimuli without contrast manipulation
	- only on-object landings (see above)
	- non-handled objects (ruler, etc.)

Binning:
	Landing positions are first binned per participant.
	
According to Françoise it's not surprising that we don't find any OVP effects, because the spread of the
landing positions is too large (the PVL effect is too strong; no saccades towards the extremes of the objects).
	
TODO: with other factor (e.g. handle side) taken into account?
TODO: bin without direction (i.e. further or closer from center, irrespective of direction)
TODO: How does Nuthmann do the binning?

8:

004B/bckp/getDM.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Filename: getDM.py

"""
DESCRIPTION:
Apply selection criteria, add useful variables, etc.

Note that I end up with four types of RT:
- 'response_time' as returned by OpenSesame -> this one is not useful because
	it doesn't take the speed of the fixation checks into account
- 'rtFromStim' is the interval between stimulus onset and response

NOTE the difference with 004: no fixation checks were applied.
	
TODO: determine cutoff criteria.
"""

# Import Python modules:
import numpy as np
import os
import sys

# Import own modules:
from exparser.DataMatrix import DataMatrix
import studies._004.constants as constants
import ascParser004B


def getDM(driftCorr, exp, excludeErrors = False, saccTooFast = 80, saccTooSlow = None,\
		rtTooFast = None, rtTooSlow = None, rtVar = 'rtFromStim', \
		exclFailedChecks = True):
	
	"""
	Applies several exclusion criteria to obtain a filtered dm.
	
	Arguments:
	
	Keyword arguments:
	excludeErrors		--- Boolean indicating whether error trials should be
							excluded. Default = False.
	saccTooFast			--- Cutoff for exclusion of too-fast saccades. Set to None
							to deactivate this filter. Default = 80.
	saccTooSlow 			--- Cutoff for exclusion of too-slow saccades. Set to None
							to deactivate this filter.
	rtTooFast			--- Cutoff for exclusion of too-fast RTs. Set to None to
							deactivate this filter.
	rtTooSlow			--- Cutoff for exclusion of too-slow RTs. Set to None to 
							deactivate this filter.
	rtVar				--- rt variable to use for excluding too fast and/or
							too slow response times.
	exclFailedChecks	--- Boolean indicating whether or not to excude trials on
							which one or more fix checks failed. Default = False
	"""

	# Declare constants:
	src = '/home/lotje/python-libs/studies/_004B'
	
	# Get the data matrix after parsing:
	dm = parseAsc.parseAsc(driftCorr = driftCorr)
	
	# Exclude Sebastiaan:
	dm = dm.select('file != "00401.asc"')
	
	# Exclude Dash because he's left handed:
	dm = dm.select('file != "00402.asc"')

	# Add a column containing driftcorr:
	dm = dm.addField("driftCorr", dtype = str)
	dm["driftCorr"] = driftCorr

	# Exclude practice trials:
	dm = dm.select("rep != 'practice'")
	
	# During parsing, tirals on which the variable 'response' did not contain
	# an int were given the value -1. Otherwise the dm will make all the 
	# 'response' values strings. Therefore, start by excluding those eventual
	# -1's:
	dm = dm.select("response != -1")
	
	# Check whether the dummy variables for the newly-determined variables
	# are all overwritten:
	
	#for dummyVar in ('saccLat', 'rtFromStim','xCoG', \
		#'yCoG', 'xCoGMask', 'yCoGMask','endX', 'endY', 'endXCorr', 'endXCorrMask'):
		#dm = dm.select('%s != %s' %(dummyVar,constants.parseDummyVar))
	
	# Saccade latencies of 0 (or -1) should not happen:
	# If no saccade is detected during a trial, 'saccLat' is changed from -1000
	# to -1 during parsing:
	dm = dm.select('saccLat > 0')
	
	# Negative new RT's should not happen:
	dm = dm.select('rtFromStim > 0')
	
	# Add values indicating whether the eyes landed towards the handle (positive
	# values) or towards the other side of the object (negative values):
	# Also, make sure all values are eventually converted to visual degrees:
	
	# Without any correction:
	dm = dm.addField("pxTowardsHandle")
	dm = dm.addField("degrTowardsHandle")
	indicesRight = np.where(dm["handle_side"] == 'right')
	indicesLeft = np.where(dm["handle_side"] == 'left')
	dm["pxTowardsHandle"][indicesRight] = (dm["endX"][indicesRight])
	dm["pxTowardsHandle"][indicesLeft] = dm["endX"][indicesRight]*-1
	dm["degrTowardsHandle"] = dm["pxTowardsHandle"]/constants.ratioPxDegr

	# With CoG correction based on the original bitmap (without mask):
	dm = dm.addField("pxTowardsHandleCorr")
	dm = dm.addField("degrTowardsHandleCorr")
	indicesRight = np.where(dm["handle_side"] == 'right')
	indicesLeft = np.where(dm["handle_side"] == 'left')
	dm["pxTowardsHandleCorr"][indicesRight] = (dm["endXCorr"][indicesRight])
	dm["pxTowardsHandleCorr"][indicesLeft] = dm["endXCorr"][indicesRight]*-1
	dm["degrTowardsHandleCorr"] = dm["pxTowardsHandleCorr"]/constants.ratioPxDegr

	# Wit CoG correction with taking mask into account::
	dm = dm.addField("pxTowardsHandleCorrMask")
	dm = dm.addField("degrTowardsHandleCorrMask")
	indicesRight = np.where(dm["handle_side"] == 'right')
	indicesLeft = np.where(dm["handle_side"] == 'left')
	dm["pxTowardsHandleCorrMask"][indicesRight] = \
		(dm["endXCorrMask"][indicesRight])
	dm["pxTowardsHandleCorrMask"][indicesLeft] = \
		dm["endXCorrMask"][indicesRight]*-1
	dm["degrTowardsHandleCorrMask"] = \
		dm["pxTowardsHandleCorrMask"]/constants.ratioPxDegr

	## Convert landing positions to visual degrees:
	#dm = dm.addField("endXDegr")
	#dm = dm.addField("endXCorrDegr")
	#dm = dm.addField("endXCorrMaskDegr")
	#dm["endXDegr"] = dm["endX"]/constants.ratioPxDegr
	#dm["endXCorrDegr"] = dm["endXCorr"]/constants.ratioPxDegr
	#dm["endXCorrMaskDegr"] = dm["endXCorrMask"]/constants.ratioPxDegr

	## Convert starting position to visual degrees:
	#dm = dm.addField("startXDegr")
	#dm["startXDegr"] = dm["startX"]/constants.ratioPxDegr
	
	#dm = dm.addField("rtFromLanding")
	#dm['rtFromLanding'] = dm['rtFromStim']-dm['saccLat']
	
	if saccTooFast != None:
		dm = dm.select('saccLat > %s'%saccTooFast)
	if saccTooSlow != None:
		dm = dm.select('saccLat < %s'%saccTooSlow)

	if rtTooFast != None:
		dm = dm.select('%s > %s'%(rtVar,rtTooFast))
	
	if rtTooSlow != None:
		dm = dm.select('%s < %s'%(rtVar,rtTooSlow))
	
	## TODO: check with graph !!
	## Add a variable containing the 'heavy side':
	#dm = dm.addField("heavy_side", dtype = str)
	#dm['heavy_side'] = "neutral"
	#dm['heavy_side'] [np.where (dm['xCoGMask'] < (0 - constants.CoGCorrTooLarge))[0]] = "left"
	#dm['heavy_side'] [np.where (dm['xCoGMask'] > (0 + constants.CoGCorrTooLarge))[0]] = "right"
	
	if excludeErrors:
		dm = dm.select('accuracy == 1')
	
	dm.save(os.path.join(src,"data - drift correction = %s - selected.csv"%driftCorr))
	
		
	return dm

if __name__ == "__main__":
	
	getDM(driftCorr = True)
	getDM(driftCorr = False)
